polynote-frontend/polynote/data/edit_buffer.ts

```diff
@@ -42,16 +50,39 @@ export class EditBuffer {
      * @param to   The end version, inclusive
      * @returns {Array}
      */
-    range(from: number, to: number): NotebookUpdate[] {
+    private rawRange(from: number, to: number): Version[] {
         let i = 0;
-        while (i < this.versions.length && this.versions[i].version <= from) {
+        while (i < this._versions.length && this._versions[i].version <= from) {
             i++;
         }
-        const edits = [];
-        while (i < this.versions.length && this.versions[i].version <= to) {
-            edits.push(...this.versions[i].edits);
+        const versions = [];
+        while (i < this._versions.length && this._versions[i].version <= to) {
+            versions.push(this._versions[i]);
             i++;
         }
-        return edits;
+        return versions;
+    }
+
+    range(from: number, to: number): NotebookUpdate[] {
+        return this.rawRange(from, to).map(ver => ver.edit)
+    }
+
+    rebaseThrough(update: NotebookUpdate, targetVersion: number): NotebookUpdate {
```

Comment:
This is the logic that's necessary to make rebasing work perfectly. When we rebase an edit from the server, we have to go through every local version since then and rebase the server's edit against that, and _also_ the that edit against the server's edit. This gives an updated server edit, and "leftovers" of the local version. Then we have to update the edit buffer and replace that local version with the "leftovers", so that the local version will be applied correctly to another future server edit.

The server does the same thing for client edits, but in the case of edits that are exactly equal, they have opposite behavior.

- [ ] Add a comment to this effect

Reply:
Another note: one thing we could do here is to take the Scala implementation of `ClientUpdateBuffer` (which is in `KernelPublisherIntegrationTest`, the thing that tests rebasing using simulated clients) and expose it with ScalaJS so that we can use it directly on the client side. I just haven't figured out how to accomplish that. I guess it would also entail sharing `NotebookUpdate` and some other data structures; I'm not sure whether that would be good or bad.